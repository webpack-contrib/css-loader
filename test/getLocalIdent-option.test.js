const path = require('path');

const { webpack, evaluated } = require('./helpers');

describe('getLocalIdent option', () => {
  it('should work (`modules: true`)', async () => {
    const config = {
      loader: {
        options: {
          modules: true,
          getLocalIdent() {
            return 'foo';
          },
        },
      },
    };
    const testId = './modules/getLocalIdent.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);
    const evaluatedModule = evaluated(module.source);

    expect(evaluatedModule).toMatchSnapshot('module (evaluated)');
    expect(evaluatedModule.locals).toMatchSnapshot('locals');
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });

  it('should work (`modules: false`)', async () => {
    const config = {
      loader: {
        options: {
          modules: false,
          getLocalIdent() {
            return 'foo';
          },
        },
      },
    };
    const testId = './modules/getLocalIdent.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);
    const evaluatedModule = evaluated(module.source);

    expect(evaluatedModule).toMatchSnapshot('module (evaluated)');
    expect(evaluatedModule.locals).toMatchSnapshot('locals');
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });

  it('should accepts arguments', async () => {
    const config = {
      loader: {
        options: {
          modules: true,
          localIdentRegExp: 'regExp',
          context: 'context',
          hashPrefix: 'hash',
          getLocalIdent(loaderContext, localIdentName, localName, options) {
            expect(loaderContext).toBeDefined();
            expect(typeof localIdentName).toBe('string');
            expect(typeof localName).toBe('string');
            expect(options).toBeDefined();

            expect(options.regExp).toBe('regExp');
            expect(options.context).toBe('context');
            expect(options.hashPrefix).toBe('hash');

            return 'foo';
          },
        },
      },
    };
    const testId = './modules/getLocalIdent.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);
    const evaluatedModule = evaluated(module.source);

    expect(evaluatedModule).toMatchSnapshot('module (evaluated)');
    expect(evaluatedModule.locals).toMatchSnapshot('locals');
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });

  it('should respect `context` option', async () => {
    const config = {
      loader: {
        options: {
          context: path.resolve(__dirname, 'fixtures/modules'),
          modules: true,
        },
      },
    };
    const testId = './modules/getLocalIdent.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);
    const evaluatedModule = evaluated(module.source);

    expect(evaluatedModule).toMatchSnapshot('module (evaluated)');
    expect(evaluatedModule.locals).toMatchSnapshot('locals');
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });
});
