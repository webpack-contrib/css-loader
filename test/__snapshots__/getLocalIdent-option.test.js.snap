// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLocalIdent option should accepts arguments: errors 1`] = `Array []`;

exports[`getLocalIdent option should accepts arguments: locals 1`] = `
Object {
  "abc": "foo",
  "def": "foo",
  "ghi": "foo",
  "jkl": "foo",
}
`;

exports[`getLocalIdent option should accepts arguments: module (evaluated) 1`] = `
Array [
  Array [
    1,
    ".foo .foo {
  color: red;
}

.foo .foo {
  color: blue;
}
",
    "",
  ],
]
`;

exports[`getLocalIdent option should accepts arguments: warnings 1`] = `Array []`;

exports[`getLocalIdent option should respect \`context\` option: errors 1`] = `Array []`;

exports[`getLocalIdent option should respect \`context\` option: locals 1`] = `
Object {
  "abc": "_1hksQTRR0UD9eKPUBlgn0X",
  "def": "_3oo37UGTAgyDe0MeQom-28",
  "ghi": "_2ZRWT_7WiIKpOei7U0eJzJ",
  "jkl": "aQ1rQfhbWSRMXFXxIfQcx",
}
`;

exports[`getLocalIdent option should respect \`context\` option: module (evaluated) 1`] = `
Array [
  Array [
    1,
    "._1hksQTRR0UD9eKPUBlgn0X ._3oo37UGTAgyDe0MeQom-28 {
  color: red;
}

._2ZRWT_7WiIKpOei7U0eJzJ .aQ1rQfhbWSRMXFXxIfQcx {
  color: blue;
}
",
    "",
  ],
]
`;

exports[`getLocalIdent option should respect \`context\` option: warnings 1`] = `Array []`;

exports[`getLocalIdent option should work (\`modules: false\`): errors 1`] = `Array []`;

exports[`getLocalIdent option should work (\`modules: false\`): locals 1`] = `
Object {
  "def": "foo",
  "ghi": "foo",
  "jkl": "foo",
}
`;

exports[`getLocalIdent option should work (\`modules: false\`): module (evaluated) 1`] = `
Array [
  Array [
    1,
    ".abc .foo {
  color: red;
}

.foo .foo {
  color: blue;
}
",
    "",
  ],
]
`;

exports[`getLocalIdent option should work (\`modules: false\`): warnings 1`] = `Array []`;

exports[`getLocalIdent option should work (\`modules: true\`): errors 1`] = `Array []`;

exports[`getLocalIdent option should work (\`modules: true\`): locals 1`] = `
Object {
  "abc": "foo",
  "def": "foo",
  "ghi": "foo",
  "jkl": "foo",
}
`;

exports[`getLocalIdent option should work (\`modules: true\`): module (evaluated) 1`] = `
Array [
  Array [
    1,
    ".foo .foo {
  color: red;
}

.foo .foo {
  color: blue;
}
",
    "",
  ],
]
`;

exports[`getLocalIdent option should work (\`modules: true\`): warnings 1`] = `Array []`;
